File: Mediation/AdMobExample/Objective-C/VponAdmobSample_objc/VponAdmobSampleObjC/AdMob/VponAdmobBannerInTableViewController.m
//
//  VponAdmobBannerInTableViewController.m
//  VponAdmobSampleObjC
//
//  Created by EricChien on 2018/12/12.
//  Copyright © 2018 Soul. All rights reserved.
//

#import "VponAdmobBannerInTableViewController.h"
#import "AdTableViewCell.h"

@import GoogleMobileAds;

#define ADPOS 5
#define DATANUM 10

#define TITLE @"title"

@interface VponAdmobBannerInTableViewController () <GADBannerViewDelegate, UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) NSMutableArray *datas;

@property (nonatomic, strong) GADBannerView *gadBannerView;

@property (nonatomic, weak) IBOutlet UITableView *mainTable;

@end

@implementation VponAdmobBannerInTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"AdMob - BannerInTable";
    
    _datas = [[NSMutableArray alloc] init];
    for (NSInteger index = 0; index < DATANUM; index++) {
        [_datas addObject:@{TITLE: [NSString stringWithFormat:@"Data %ld", (long)index]}];
    }
    
    [self requestButtonDidTouch:nil];
}

#pragma mark - Button Method

- (IBAction) requestButtonDidTouch:(UIButton *)sender {
    if (_gadBannerView.superview) {
        [_gadBannerView removeFromSuperview];
    }
    
    GADRequest *request = [GADRequest request];
//    GADExtras *extra = [[GADExtras alloc] init];
//    extra.additionalParameters = @{
//        @"contentURL": @"https://www.vpon.com",
//        @"contentData": @{@"key1": @"Admob", @"key2": @(1.2), @"key3": @(YES)},
//        @"friendlyObstructions": @[@{ @"view": [[UIView alloc] init], @"purpose": @(2), @"desc": @"not_visible"}]
//    };
//    [request registerAdNetworkExtras:extra];
//    request.testDevices = @[kGADSimulatorID];
    
    _gadBannerView = [[GADBannerView alloc] initWithAdSize:GADAdSizeMediumRectangle];
    _gadBannerView.adUnitID = @"";
    _gadBannerView.delegate = self;
    _gadBannerView.rootViewController = self;
    [_gadBannerView loadRequest:request];
}

#pragma mark - GADBannerView Delegate

- (void)bannerViewDidReceiveAd:(nonnull GADBannerView *)bannerView {
    NSLog(@"Received banner ad successfully");
    [_mainTable reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:ADPOS inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
}

- (void)bannerView:(nonnull GADBannerView *)bannerView didFailToReceiveAdWithError:(nonnull NSError *)error {
    NSLog(@"Failed to receive banner with error: %@", [error localizedFailureReason]);
    [_mainTable reloadRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:ADPOS inSection:0]] withRowAnimation:UITableViewRowAnimationNone];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _datas.count + 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == ADPOS) {
        AdTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"AdTableViewCell" forIndexPath:indexPath];
        if (_gadBannerView) {
            [cell.contentView addSubview:_gadBannerView];
            _gadBannerView.center = cell.contentView.center;
        }
        return cell;
    } else {
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"reuseIdentifier" forIndexPath:indexPath];
        NSIndexPath *current = [self currentIndexPath:indexPath];
        NSDictionary *data = [_datas objectAtIndex:current.row];
        cell.textLabel.text = [NSString stringWithFormat:@"%@", data[TITLE]];
        return cell;
    }
}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row == ADPOS) {
        return 250;
    } else {
        return 44;
    }
}

- (NSIndexPath *) currentIndexPath:(NSIndexPath *)indexPath {
    if (indexPath.row < ADPOS) {
        return indexPath;
    } else {
        return [NSIndexPath indexPathForRow:indexPath.row - 1 inSection:indexPath.section];
    }
}

@end

File: Mediation/AdMobExample/Swift/VponAdmobSample_swift/VponAdmobSampleSwift/AdMob/VponAdmobBannerInTableViewController.swift
//
//  VponAdmobBannerInTableViewController.swift
//  VponAdmobSampleSwift
//
//  Created by Yi-Hsiang, Chien on 2020/9/25.
//  Copyright © 2020 Soul. All rights reserved.
//

import UIKit
import GoogleMobileAds

extension VponAdmobBannerInTableViewController: UITableViewDataSource, UITableViewDelegate {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        if indexPath.row == adPos {
            let cell = tableView.dequeueReusableCell(withIdentifier: "adReuseIdentifier", for: indexPath) as! AdTableViewCell
            if let banner = gadBannerView {
                cell.contentView.addSubview(banner)
                banner.center = cell.contentView.center
            }
            return cell
        } else {
            let cell = tableView.dequeueReusableCell(withIdentifier: "reuseIdentifier", for: indexPath)
            cell.textLabel?.text = "Data \(indexPath.row + 1)"
            return cell
        }
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return indexPath.row == adPos ? 250 : 44
    }
}

extension VponAdmobBannerInTableViewController: GADBannerViewDelegate {
    func bannerViewDidReceiveAd(_ bannerView: GADBannerView) {
        print("Received banner ad successfully")
        mainTable.reloadRows(at: [IndexPath.init(row: adPos, section: 0)], with: .none)
    }
    
    func bannerView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: Error) {
        print("Failed to receive banner with error: \(error.localizedDescription)")
    }
}

class VponAdmobBannerInTableViewController: UIViewController {
    
    @IBOutlet var mainTable: UITableView!
    
    var gadBannerView: GADBannerView?

    let adPos = 5
    let items = 20
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        addBanner()
    }
    
    func addBanner() {
        let request = GADRequest()
        let extra = GADExtras()
        extra.additionalParameters = [
            "contentURL":"https://www.vpon.com",
            "contentData": ["key1": "Admob", "key2": 1.2, "key3": true],
            "friendlyObstructions": [["view": UIView(), "purpose": 2, "desc": "not_visible"]]
        ]
        request.register(extra)
        
        gadBannerView = GADBannerView(adSize: GADAdSizeMediumRectangle)
// TODO: set ad unit id
        gadBannerView!.adUnitID = ""
        gadBannerView!.delegate = self
        gadBannerView!.rootViewController = self
        gadBannerView!.load(request)
    }
}
